lib/st/api/lims/ml_warehouse.pm

```diff
@@ -32,6 +33,39 @@ in WTSI::DNAP::Warehouse::Schema.
 
 =head2 id_flowcell_lims
 
+=head2 id_run
+
+id_run, optional attribute.
+
+=cut
+has '+id_run' =>       ( required        => 0,
+                         lazy_build      => 1,
+                       );
+
+sub _build_id_run {
+  my $self = shift;
+  if (not $self->has_flowcell_barcode and not $self->has_id_flowcell_lims) {
+    croak 'Require flowcell_barcode or id_flowcell_lims to try to find id_run';
+  }
+  my%search;
+  if($self->has_flowcell_barcode){
+    $search{'flowcell_barcode'} = $self->flowcell_barcode;
+  }
+  if($self->has_id_flowcell_lims){
+    $search{'id_flowcell_lims'} = $self->id_flowcell_lims;
+  }
+  my$id_run_rs=$self->iseq_flowcell->related_resultset('iseq_product_metrics')->search(\%search,{'columns'=>[qw(id_run)], 'distinct'=>1});
+  my$count = $id_run_rs->count;
+  croak "Found more than one ($count) id_run" if ($count > 1);
+  if($count){
+    return $id_run_rs->first->id_run;
+  }
+  carp join q( ), 'No id_run set yet',
+    ($self->has_flowcell_barcode ? ('flowcell_barcode:'.$self->flowcell_barcode) : ()),
+    ($self->has_id_flowcell_lims ? ('id_flowcell_lims:'.$self->id_flowcell_lims) : ());
+  return;
```

Comment:
A test is needed for this case. id_run is an int, so it I expect an error with undef return. Is this the intension? If error is OK, then it might be better to error explicitly, ie replace carp with croak.